### ISSUE TITLE
Contrasting contrastive loss functions.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#57]
Zichen Wang. [Contrasting contrastive loss functions](https://towardsdatascience.com/contrasting-contrastive-loss-functions-3c13ca5f055e).

**Comment by NorbertZheng** [comment]
## Overview
A comprehensive guide to four contrastive loss functions for contrastive learning.

[In a previous post](https://towardsdatascience.com/contrastive-loss-for-supervised-classification-224ae35692e7), I wrote about contrastive learning in supervised classification and performed some experiments on MNIST dataset and alike to find that the two-stage method proposed in the [Khosla et al. 2020 paper](https://arxiv.org/abs/2004.11362) indeed shows significant improvement for supervised classification task by learning meaningful embeddings with contrastive loss. Later I found my experiments actually used a different contrastive loss function than Khosla et al. proposed. Although sharing the same intuition of explicitly contrasting examples against each other with respect to their labels, different contrastive loss functions can have their own nuances. In this post, I will review a series of contrastive loss functions and compare their performances in supervised classification task.

**Comment by NorbertZheng** [comment]
## Preliminary
Contrastive loss functions were invented for [metric learning](https://en.wikipedia.org/wiki/Similarity_learning), which intends to
 - learn similarity functions that measure the similarity or distance between a pair of objects.

In the context of classification, the desired metric would render a pair of examples with the same label more similar than a pair of examples with different labels. **Deep metric learning** involves deep neural networks to embed data points to a lower-dimensional space with nonlinearity, then uses contrastive loss function to optimize the parameters in the neural networks. Recent research projects have applied deep metric learning to self-supervised learning, supervised learning and even reinforcement learning, for example, [Contrastively-trained Structured World Models (C-SWMs)](https://arxiv.org/abs/1911.12247).

To review different contrastive loss functions in the context of deep metric learning, I use the following formalization. Let $x$ be the input feature vector and $y$ be its label. Let $f(\cdot)$ be a encoder network mapping the input space to the embedding space and let $z=f(x)$ be the embedding vector.

**Comment by NorbertZheng** [comment]
Max margin contrastive loss should already be enough to train PEs of TEM alone.

**Comment by NorbertZheng** [comment]
### Multi-class N-pair loss (Sohn 2016)
Multi-class N-pair loss is a generalization of triplet loss allowing joint comparison among **more than one negative samples**. When applied on a pair of positive samples $z_{i}$ and $z_{j}$ sharing the same label $(y_{i}=y_{j})$ from a mini-batch with $2N$ samples, it is calculated as:
![image](https://user-images.githubusercontent.com/37235023/216612390-38893dd0-a722-4418-966d-4c4748c64b9b.png)
where $z_{i}z_{j}$ is the inner product, which is equivalent to cosine similarity when both vectors have unit norm.

As the figure below shows, $N$-pair loss pushes $2(N-1)$ negative samples away simultaneously instead of one at a time:
![image](https://user-images.githubusercontent.com/37235023/216612686-6f8d17a2-915f-45f2-98c2-44f052189efa.png)
Triplet loss (left) and its extension $(N+1)$-triplet loss (right) (from Sohn 2016)

With some algebraic manipulations, multi-class $N$-pair loss can be written as the following:
![image](https://user-images.githubusercontent.com/37235023/216612901-e17df8b3-84a1-42dd-8cab-0326df5e3cfd.png)
This form of multi-class $N$-pair loss helps us introduce the next loss function.

**Comment by NorbertZheng** [comment]
### Supervised NT-Xent loss (Khosla et al. 2020)
Let’s first look at the self-supervised version of NT-Xent loss. NT-Xent is coined by [Chen et al. 2020 in the SimCLR paper](https://arxiv.org/abs/2002.05709) and is short for “normalized temperature-scaled cross-entropy loss”. It is a modification of the multi-class N-pair loss with the addition of the temperature parameter $(\tau)$ to scale the cosine similarities:
![image](https://user-images.githubusercontent.com/37235023/216613230-de4a3d31-d39a-480c-baa0-590465f5ea93.png)
Self-supervised NT-Xent loss.

Chen et al. found that an appropriate temperature parameter can help the model learn from hard negatives. In addition, they showed that the optimal temperature differs in different
 - batch sizes,
 - number of training epochs.

Khosla et al. later extended NT-Xent loss for supervised learning:
![image](https://user-images.githubusercontent.com/37235023/216613615-42b98452-3942-48fa-b575-d7a4bed705d8.png)
Supervised NT-Xent loss.

**Comment by NorbertZheng** [comment]
## Experimental results
Next I assess the whether these contrastive loss functions can help the encoder network to learn meaningful representations of the data to aid the classification task. Following the exact same experimental settings from [my previous post](https://docs.google.com/presentation/d/1iB59aKvWtjeN2ZYlPih2ygH8R4J8YS6k3iGQJNfukeE/edit?usp=sharing), with small batch size (32) and low learning rate (0.001), I found
 - all these contrastive loss functions **except for triplet loss with hard negative mining** outperforms the MLP baseline without the stage 1 pre-training:

![image](https://user-images.githubusercontent.com/37235023/216614583-87a60eb9-28d9-4813-ab2e-ba3edd211696.png)
Performance (accuracy) on the hold-out test sets of MNIST and Fashion MNIST datasets (Results from triplet with hard negative mining not shown).

These results confirmed the benefit of using contrastive loss function in the pre-training the encoder part of the network for the subsequent classification. It also underscored the importance of triplet mining for triplet loss. Specifically,
 - **semi-hard mining** works the best on these experiments, which is consistent with the FaceNet paper.

 - Both Chen et al. (SimCLR) and Khosla et al. use very large batch sizes and higher learning rates for the NT-Xent loss to achieve greater performances.

I next experimented with different batch sizes 32, 256 and 2048 with learning rates of 0.001, 0.01, and 0.2, respectively.
![image](https://user-images.githubusercontent.com/37235023/216615023-fc0be45c-5c21-4765-b22b-564449bf11ac.png)

The results show that the performances diminish as the batch size increases for all loss functions. Although triplet loss with semi-hard negative mining performs very well on small/medium batches, it is very memory intensive and my 16G RAM is impossible to handle it with a batch size of 2048. Supervised NT-Xent loss does turn to perform relatively better on larger batch size compared to its counterparts.
 - There could be space for improvement for supervised NT-Xent if I were to optimize the temperature parameter. The temperature I used was 0.5.

Next, I checked the PCA projections of the embeddings learned using contrastive loss functions to see if they learn any informative representations during the pre-training stage.
![image](https://user-images.githubusercontent.com/37235023/216615511-2c6a98b9-0d50-47d5-a89f-d3cdda238705.png)
PCA projections of the embeddings learned by encoder networks with different contrastive loss functions and batch sizes on MNIST dataset. From left to right: projections learned by 1) max margin loss; 2) triplet loss with semi-hard mining; 3) multi-class N-pair loss; 4) supervised NT-Xent loss. From top to bottom: batch sizes of 32, 256, 2048.

![image](https://user-images.githubusercontent.com/37235023/216615661-3edae913-e181-4066-a626-22580ca19588.png)
Joint plots showing the densities in the PCA projections learned by models on MNIST dataset. From left to right: projections learned by 1) max margin loss; 2) triplet loss with semi-hard mining; 3) multi-class N-pair loss; 4) supervised NT-Xent loss. From top to bottom: batch sizes of 32, 256, 2048.

Judging from the colored PCA projections and their densities, we can see both max margin and supervised NT-Xent learn tighter clusters for each class, whereas clusters from triplet loss with semi-hard mining are most dilated but still distinctive. As the batch size increase, the representation qualities degenerate in multi-class $N$-pair loss and max-margin loss, but
 - not so much in supervised NT-Xent loss, suggesting this loss is indeed more robust to larger batch size.

Below are the PCA projections of the learned representation on a more difficult Fashion MNIST dataset. Overall it shows similar observations with MNIST.
![image](https://user-images.githubusercontent.com/37235023/216615961-50319890-fec2-4f07-9561-cb6e8046c763.png)
PCA projections of the embeddings learned by encoder networks with different contrastive loss functions and batch sizes on Fashion MNIST dataset. From left to right: projections learned by 1) max margin loss; 2) triplet loss with semi-hard mining; 3) multi-class N-pair loss; 4) supervised NT-Xent loss. From top to bottom: batch sizes of 32, 256, 2048.

**Comment by NorbertZheng** [comment]
## Code
Codes used for these experiments are available here: https://github.com/wangz10/contrastive_loss.

**Comment by NorbertZheng** [comment]
## References
 - [Hadsell, R., Chopra, S., & LeCun, Y. (2006, June). Dimensionality reduction by learning an invariant mapping.](http://yann.lecun.com/exdb/publis/pdf/hadsell-chopra-lecun-06.pdf) In 2006 IEEE Computer Society Conference on Computer Vision and Pattern Recognition (CVPR’06) (Vol. 2, pp. 1735–1742). IEEE.
 - [Weinberger, K. Q., Blitzer, J., & Saul, L. K. (2006). Distance metric learning for large margin nearest neighbor classification.](https://papers.nips.cc/paper/2795-distance-metric-learning-for-large-margin-nearest-neighbor-classification.pdf) In Advances in neural information processing systems (pp. 1473–1480).
 - [Schroff, F., Kalenichenko, D., & Philbin, J. (2015). Facenet: A unified embedding for face recognition and clustering.](https://arxiv.org/abs/1503.03832) In Proceedings of the IEEE conference on computer vision and pattern recognition (pp. 815–823).
 - [Sohn, K. (2016). Improved deep metric learning with multi-class n-pair loss objective.](https://papers.nips.cc/paper/6200-improved-deep-metric-learning-with-multi-class-n-pair-loss-objective) In Advances in neural information processing systems (pp. 1857–1865).
 - [Chen, T., Kornblith, S., Norouzi, M., & Hinton, G. (2020). A simple framework for contrastive learning of visual representations.](https://arxiv.org/pdf/2002.05709.pdf) arXiv preprint arXiv:2002.05709.
 - [Khosla, P., Teterwak, P., Wang, C., Sarna, A., Tian, Y., Isola, P., … & Krishnan, D. (2020). Supervised Contrastive Learning.](https://arxiv.org/pdf/2004.11362.pdf) arXiv preprint arXiv:2004.11362.

**Comment by wicaksonoleksono** [comment]
make it a paper pls, need to cite this
